sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala

```diff
@@ -103,13 +103,13 @@ object ExternalCatalogUtils {
     }
     var plaintextEndIdx = path.indexOf('%')
     val length = path.length
-    if (plaintextEndIdx == -1 || plaintextEndIdx + 2 > length) {
+    if (plaintextEndIdx == -1 || plaintextEndIdx + 2 >= length) {
       // fast path, no %xx encoding found then return the string identity
       path
     } else {
       val sb = new java.lang.StringBuilder(length)
       var plaintextStartIdx = 0
-      while(plaintextEndIdx != -1 && plaintextEndIdx + 2 < length) {
+      do {
```

Comment:
FYI, for scala 3, do-while is removed

Reply:
Yes, what @yaooqinn  said is correct:
```scala
Welcome to Scala 3.3.1 (17.0.11, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                                                                                                                                                                                     
scala> var x = 10
     | 
     | do {
     |   println("Value of x: " + x)
     |   x = x - 1
     | } while (x > 0)
-- Error: ----------------------------------------------------------------------
3 |do {
  |^^
  |end of statement expected but 'do' found

```

Reply:
Thank you for the remind. I will revert it.